core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java

```diff
@@ -0,0 +1,735 @@
+//import static org.elasticsearch.index.query.RangeQueryBuilder.TIME_ZONE_FIELD;
+/** A {@link FieldMapper} for indexing numeric and date ranges, and creating queries */
+public class RangeFieldMapper extends FieldMapper {
+    public static final boolean DEFAULT_INCLUDE_UPPER = true;
+    public static final boolean DEFAULT_INCLUDE_LOWER = true;
+
+    public static class Defaults {
+        public static final Explicit<Boolean> COERCE = new Explicit<>(true, false);
+    }
+
+    // this is private since it has a different default
+    static final Setting<Boolean> COERCE_SETTING =
+        Setting.boolSetting("index.mapping.coerce", true, Setting.Property.IndexScope);
+
+    public static class Builder extends FieldMapper.Builder<Builder, RangeFieldMapper> {
+        private Boolean coerce;
+        private Locale locale;
+
+        public Builder(String name, RangeType type) {
+            super(name, new RangeFieldType(type), new RangeFieldType(type));
+            builder = this;
+            locale = Locale.ROOT;
+        }
+
+        @Override
+        public RangeFieldType fieldType() {
+            return (RangeFieldType)fieldType;
+        }
+
+        public Builder coerce(boolean coerce) {
+            this.coerce = coerce;
+            return builder;
+        }
+
+        protected Explicit<Boolean> coerce(BuilderContext context) {
+            if (coerce != null) {
+                return new Explicit<>(coerce, true);
+            }
+            if (context.indexSettings() != null) {
+                return new Explicit<>(COERCE_SETTING.get(context.indexSettings()), false);
+            }
+            return Defaults.COERCE;
+        }
+
+        public Builder dateTimeFormatter(FormatDateTimeFormatter dateTimeFormatter) {
+            fieldType().setDateTimeFormatter(dateTimeFormatter);
+            return this;
+        }
+
+        @Override
+        public Builder nullValue(Object nullValue) {
+            throw new IllegalArgumentException("Field [" + name() + "] does not support null value.");
+        }
+
+        public void locale(Locale locale) {
+            this.locale = locale;
+        }
+
+        @Override
+        protected void setupFieldType(BuilderContext context) {
+            super.setupFieldType(context);
+            FormatDateTimeFormatter dateTimeFormatter = fieldType().dateTimeFormatter;
+            if (!locale.equals(dateTimeFormatter.locale())) {
+                fieldType().setDateTimeFormatter(new FormatDateTimeFormatter(dateTimeFormatter.format(),
+                    dateTimeFormatter.parser(), dateTimeFormatter.printer(), locale));
+            }
+        }
+
+        @Override
+        public RangeFieldMapper build(BuilderContext context) {
+            setupFieldType(context);
+            return new RangeFieldMapper(name, fieldType, defaultFieldType, coerce(context), includeInAll,
+                context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo);
+        }
```

Comment:
Since doc values are not supported yet, I think we'd need something that fails if a user tries to enable them?